text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,338 @@
+</template>
+```
+
+Imperatively:
+
+```javascript
+import { LightningElement } from "lwc";
+
+export default class DynamicCtor extends LightningElement {
+  customCtor;
+
+  connectedCallback() {
+    this.loadCtor();
+  }
+
+  async loadCtor() {
+    const { default: ctor } = await import("c/customConstructor");
+    this.customCtor = ctor;
+    this.template.querySelector(ctor.lightningElementName).setAttribute('title', 'hawaii');
+    this.template.querySelector(ctor.lightningElementName).addEventListener('click', () => console.log('hello world'));
+  }
+}
+```
+#### Usage with other LWC directives
+
+All [LWC custom element directives](https://lwc.dev/guide/reference#directives-for-nested-templates) with the exception of `lwc:external` are available to use on `<lwc:dynamic>` and their functionality will be passed through to the custom element once it has been created.
+
+`lwc:external` will not be available because the constructor provided to the `lwc:is` directive must be a `LightningElementConstructor`.
+
+#### Children of the dynamically created element
+
+The `<lwc:dynamic>` tag accepts child elements, first creating the dynamic component and subsequently the children afterwards.  Each time the dynamic component changes, the existing dynamic component will be unmounted along with all of its children.  The new dynamic component will be created along with the children.
+
+```html
+<template>
+    <lwc:dynamic lwc:is={ctor}>
+        <span>child</span>
+    </lwc:dynamic>
+</template>
+```
+
+#### Light DOM vs shadow DOM
+
+A dynamic component can switch between light DOM and shadow DOM and should follow the same semantics as currently exists with other forms of dynamic rendering components.
+
+For example
+
+```html
+<template>
+    <template lwc:if={renderLight}>
+        <light-component></light-component>
+    <template lwc:else>
+        <shadow-component></shadow-component>
+    </template>
+</template>
+```
+
+Is semantically the same as
+
+```html
+<template>
+    <!-- The ctor swaps between light and shadow DOM components -->
+    <lwc:dynamic lwc:is={ctor}></lwc:dynamic>
+</template>
+```
+
+In both cases the dynamic component will be fully mounted and unmounted.
+
+#### Styles
+
+Something to keep in mind is that [light DOM styles are injected to the closest root node](https://rfcs.lwc.dev/rfcs/lwc/0115-light-dom#styles) and are not removed once the components are unmounted. This means that repeated mounting and mounting of a dynamic component could cause styles to be overwritten.  This can be avoided by using either scoped styles or shadow dom, as the styles will be scoped.
+
+## Drawbacks
+
+The main drawback of this design is that when the constructor changes, the entire component hierarchy will need to be re-rendered. This means that the children of the custom element will also need to re-rendered.  Careful consideration should be taken by the component author as this may cause performance issues.
```

Comment:
This is only true if synthetic shadow isn't it? Let's chat more about this.

Reply:
ping

Reply:
I think when the dynamic component has a `<slot>` then there shouldn't be a difference between synthetic and native shadow.  The dynamic component will be unmounted along with its children and recreated using the new constructor.  Here's an [example](https://stackblitz.com/edit/salesforce-lwc-bsehiw?file=src/modules/x/child2/child2.js) using the current `lwc:dynamic` implementation.

If the dynamic component doesn't have a `<slot>` though there will be a difference as I believe in synthetic shadow the children will not be rendered but in native shadow they will be rendered.

This means that in synthetic shadow there are no children to mount/unmount while in native there will be.  

I'll do some more testing and let you know what I find.

Reply:
I did some testing, here's an [example](https://stackblitz.com/edit/salesforce-lwc-bsehiw?file=src/modules/x/app/app.html) comparing both native and synthetic shadow when there is no `<slot>`.

The synthetic shadow does not mount any children while the native shadow does mount the children.